### PR TITLE
CoAT changes to support Kimono build

### PR DESCRIPTION
**Pull request by johnxwork**
1. Removed most java dependencies on content/shell/android/ resources;
2. Support dynamically adding "cobalt." java package prefix when invoking JNI;
3. Avoid using hash names in JNI templates.

go/kimono-chrobalt-build
b/372559266

### CODE REVIEW

File: build/config/android/config.gni [hunk 1]
```diff
@@ -287,7 +287,8 @@ if (is_android || is_chromeos) {
     enable_incremental_d8 = true
 
     # Use hashed symbol names to reduce JNI symbol overhead.
-    use_hashed_jni_names = !is_java_debug
+    # Cobalt avoids hash names to facilitate Kimono build.
+    use_hashed_jni_names = false
```
Comment (kaidokert) [inline build/config/android/config.gni:291]: perhaps leave it `use_hashed_jni_names = !is_java_debug || is_cobalt` to ensure Chromium build configs don't change here.

File: cobalt/android/BUILD.gn [hunk 2]
```diff
@@ -44,7 +45,8 @@ android_library("cobalt_apk_java") {
     "//components/embedder_support/android:view_java",
     "//content/public/android:content_java",
     "//content/shell/android:content_shell_apk_java",
-    "//content/shell/android:content_shell_apk_resources",
+
+    #"//content/shell/android:content_shell_apk_resources",
```
Comment (kaidokert) [inline cobalt/android/BUILD.gn:49]: Extract this out down below to line 59:

```
if(!is_cobalt) {
  deps += [
     "//content/shell/android:content_shell_apk_resources"
  ]
}
```

Comment (kaidokert) [reply cobalt/android/BUILD.gn:49]: Actually nevermind - this only gets compiled _for_ cobalt configs, correct ? So it can just be removed

File: content/shell/android/shell_apk/src/org/chromium/content_shell_apk/ContentShellActivity.java [hunk 3]
```diff
@@ -21,8 +21,8 @@
 import org.chromium.content_public.browser.WebContents;
 import org.chromium.content_shell.Shell;
 import org.chromium.content_shell.ShellManager;
-import org.chromium.ui.base.ActivityWindowAndroid;
-import org.chromium.ui.base.IntentRequestTracker;
+// import org.chromium.ui.base.ActivityWindowAndroid;
```
Comment (kaidokert) [inline content/shell/android/shell_apk/src/org/chromium/content_shell_apk/ContentShellActivity.java:24]: Why are any of those changes needed in ContentShellActivity ? We cloned it to CobaltActivity already - didnt we ?

File: ui/android/BUILD.gn [hunk 4]
```diff
@@ -240,13 +240,23 @@ android_library("ui_utils_java") {
     "java/src/org/chromium/ui/KeyboardVisibilityDelegate.java",
     "java/src/org/chromium/ui/MotionEventUtils.java",
     "java/src/org/chromium/ui/UiUtils.java",
+    "java/src/org/chromium/ui/display/DisplaySwitches.java",
+    "java/src/org/chromium/ui/display/DisplayAndroid.java",
+    "java/src/org/chromium/ui/display/DisplayAndroidManager.java",
+    "java/src/org/chromium/ui/display/PhysicalDisplayAndroid.java",
   ]
```
Comment (kaidokert) [inline ui/android/BUILD.gn:247]: Factor it out so that upstream / Chromium build configs don't change

On line 247
```
if(is_cobalt) {
 sources += [
   ...
 ]
}
```

File: build/config/android/config.gni (hunk 1 above)
Comment (johnxwork) [reply build/config/android/config.gni:291]: This PR pretty much just provides two gn flags to work with Kimono build. I'm just sharing my workspace with Andrew to figure out some build problems for now. This PR is far from review yet.

File: base/android/jni_android.cc [hunk 5]
```diff
@@ -30,9 +33,48 @@ ABSL_CONST_INIT thread_local void* stack_frame_pointer = nullptr;
...
+const char* COBALT_ORG_CHROMIUM = "cobalt/org/chromium";
+const char* ORG_CHROMIUM = "org/chromium";
+
+bool g_add_cobalt_prefix = false;
+std::atomic<bool> g_checked_command_line(false);
+
+std::string getRepackagedName(const char* signature) {
+  std::string holder(signature);
+  size_t pos = 0;
+  while ((pos = holder.find(ORG_CHROMIUM, pos)) != std::string::npos) {
```
Comment (zhongqiliang) [inline base/android/jni_android.cc:46]: what does the package name look like? will there be multiple "org/chromium" in the name? if not, should we use `if` here instead of `while`?

Comment (johnxwork) [reply base/android/jni_android.cc:46]: The JNI signature can have multiple "org/chromium", such as: "function_1(org/chromium/.../param1, org/chromium/.../param2)"

File: cobalt/android/apk/app/src/main/java/dev/cobalt/coat/CobaltActivity.java [hunk 6]
```diff
@@ -97,12 +96,13 @@ public abstract class CobaltActivity extends Activity {
   private Intent mLastSentIntent;
   private String mStartupUrl;
   private IntentRequestTracker mIntentRequestTracker;
+  protected Boolean shouldSetJNIPrefix = false;
```
Comment (zhongqiliang) [inline cobalt/android/apk/app/src/main/java/dev/cobalt/coat/CobaltActivity.java:99]: I do not see this value is set to true anywhere, how does it work?

File: cobalt/android/apk/app/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -1,33 +1,26 @@
...
-  <!-- Toast message when we can't get the device account to sign in. -->
-  <string name="starboard_account_picker_error">Account not available.</string>
-  <!-- Toast message when we the account can't be authorized. -->
-  <string name="starboard_account_auth_error">Account authorization failed.</string>
-  <!-- Toast message indicating that the selected account is no longer available to sign-in. -->
-  <string name="starboard_missing_account">Account not available:\n%1$s</string>
-  <!-- Toast message shown when the user has denied permission to get device accounts. -->
-  <string name="starboard_accounts_permission">Sign in requires Contacts permission to be granted in system settings.</string>
+    <string name="url_hint">Type URL Here</string>
+    <string name="browser_process_initialization_failed">Initialization failed.</string>
```
[the remark this answers is not in the document]
Comment (johnxwork) [reply cobalt/android/apk/app/src/main/res/values/strings.xml:11]: line 247 in CobaltActivity.java

File: cobalt/android/apk/app/src/main/java/dev/cobalt/coat/CobaltActivity.java (hunk 6 above)
Comment (johnxwork) [reply cobalt/android/apk/app/src/main/java/dev/cobalt/coat/CobaltActivity.java:99]: This is a protected variable and set in child class. Check out cl/712090387

File: cobalt/android/BUILD.gn [hunk 8]
```diff
@@ -36,14 +37,81 @@ generate_jni("jni_headers") {
...
+    "apk/app/src/org/chromium/content_shell/Shell.java",
+    "apk/app/src/org/chromium/content_shell/ShellManager.java",
+  ]
+}
+
+android_library("content_shell_apk_java") {
   testonly = true
-  resources_package = "dev.cobalt.coat"
+
+  #srcjar_deps = [ ":content_javatests_aidl" ]
```
Comment (andrewsavage1) [inline cobalt/android/BUILD.gn:50]: cruft?

File: cobalt/android/BUILD.gn [hunk 9]
```diff
@@ -36,14 +37,81 @@ generate_jni("jni_headers") {
...
+  ]
+}
+
+android_library("cobalt_shell_java") {
+  testonly = true
+  resources_package = "org.chromium.content_shell"
+  deps = [
+    ":cobalt_java_resources",
+
+    #":content_shell_manifest",
```
Comment (andrewsavage1) [inline cobalt/android/BUILD.gn:81]: cruft?

File: cobalt/android/BUILD.gn [hunk 10]
```diff
@@ -120,6 +182,21 @@ android_library("cobalt_apk_java") {
     "apk/app/src/main/java/dev/cobalt/util/SystemPropertiesHelper.java",
     "apk/app/src/main/java/dev/cobalt/util/UsedByNative.java",
   ]
+
+  annotation_processor_deps = [ "//base/android/jni_generator:jni_processor" ]
+}
+
+# TODO(cobalt): Re-enable or remove disabled java files.
```
Comment (andrewsavage1) [inline cobalt/android/BUILD.gn:189]: seems obsolete

File: base/android/jni_android.cc [hunk 11]
```diff
@@ -30,9 +33,48 @@ ABSL_CONST_INIT thread_local void* stack_frame_pointer = nullptr;
...
+#endif
+
 ScopedJavaLocalRef<jclass> GetClassInternal(JNIEnv* env,
+#if BUILDFLAG(IS_COBALT)
+                                            const char* original_class_name,
+                                            jobject class_loader) {
+  const char* class_name;
+  std::string holder;
+  if (shouldAddCobaltPrefix()) {
+    holder = getRepackagedName(original_class_name);
```
Comment (andrewsavage1) [inline base/android/jni_android.cc:69]: Is this happening at runtime? Do we have to worry about a performance hit of doing this every(?) function call?

**Review by andrewsavage1 — COMMENTED** [review]
It seems to me that this PR is doing two things—moving android's cobalt_apk to depend on content instead of content_shell and then renaming the chromium package. Ideally we wouldn't do these things together, maybe we can make the PR two commits and later do a merge instead of a squash into main?

File: base/android/jni_android.cc (hunk 11 above)
Comment (johnxwork) [reply base/android/jni_android.cc:69]: Yes you are correct. I discussed this here: https://docs.google.com/document/d/1i9vJEFyloPdeUaL8NXOsCKLeVWo45KF5cxdUWb4GEiU/edit?tab=t.0#bookmark=id.i2glpx9ni220

TLDR there's a performance hit but it's not executed as frequently as I thought and we should be fine. There is room for improvement though.

**Comment by johnxwork** [comment]
> It seems to me that this PR is doing two things—moving android's cobalt_apk to depend on content instead of content_shell and then renaming the chromium package. Ideally we wouldn't do these things together, maybe we can make the PR two commits and later do a merge instead of a squash into main?

I separated the PR into two commits so that you can see the changes I made to Shell.java and ShellManager.java.

**Comment by zhongqiliang** [comment]
The meeting 's conclusion for this.
Although not ideal and incurring some technical debt, this approach provides a working solution for now.

So LGTM.

**Review by zhongqiliang — APPROVED** [review]
LGTM

File: base/android/jni_android.cc [hunk 12]
```diff
@@ -30,9 +33,48 @@ ABSL_CONST_INIT thread_local void* stack_frame_pointer = nullptr;
 
 bool g_fatal_exception_occurred = false;
 
+#if BUILDFLAG(IS_COBALT)
```
Comment (kaidokert) [inline base/android/jni_android.cc:36]: Given this lives in upstream code, i think it's probably a good place to add maybe a two-three line comment that explains what's being done and why.

E.g. something like 

```
/* Cobalt specific hack to move Java classes to a custom namespace. 
  For every class org.chromium.foo moves them to cobalt.org.chromium.foo
```

File: cobalt/android/BUILD.gn [hunk 13]
```diff
@@ -36,14 +37,77 @@ generate_jni("jni_headers") {
...
-# TODO(cobalt): Re-enable or remove disabled java files.
-android_library("cobalt_apk_java") {
+generate_jni("content_shell_jni_headers") {
+  sources = [
+    "apk/app/src/org/chromium/content_shell/Shell.java",
+    "apk/app/src/org/chromium/content_shell/ShellManager.java",
+  ]
+}
+
+android_library("content_shell_apk_java") {
```
Comment (kaidokert) [inline cobalt/android/BUILD.gn:47]: Just a note, we should evolve to a place where we can eventually get back to using `/shell/android/` content_shell_apk_java` target unmodified. But as an increment here, this is fine.

**Review by kaidokert — APPROVED** [review]
I'm okay with getting this landed

File: cobalt/android/BUILD.gn (hunk 13 above)
Comment (johnxwork) [reply cobalt/android/BUILD.gn:47]: I'd advocate for removing all content_shell deps instead and evolve these forked files to be completely different from the original version.

File: base/android/jni_android.cc [hunk 14]
```diff
@@ -30,9 +33,52 @@ ABSL_CONST_INIT thread_local void* stack_frame_pointer = nullptr;
 
 bool g_fatal_exception_occurred = false;
 
+/* Cobalt specific hack to move Java classes to a custom namespace.
```
Comment (andrewsavage1) [inline base/android/jni_android.cc:36]: fwiw these could be made static inline functions in a cobalt-owned helpers file instead, but I'll leave that to your discretion

**Review by andrewsavage1 — APPROVED** [review]
